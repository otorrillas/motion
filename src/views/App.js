import React, { PureComponent, Suspense } from 'react';

import styles from './App.module.css';

import { fetchTrackList } from '../api';
import { PlayerContext } from './PlayerContext';

import Loader from '../components/Loader';
import TrackList from '../components/TrackList';
import Player from '../components/Player';

class App extends PureComponent {
  state = {
    trackList: null,
    currentTrack: null,
    isPlaying: false,
    playerRef: React.createRef()
  };

  componentDidMount() {
    this._loadAsyncData();
  }

  componentWillUnmount() {
    if (this._asyncRequest) {
      this._asyncRequest.cancel();
    }
  }

  _loadAsyncData() {
    this._asyncRequest = fetchTrackList().then(trackList => {
      this._asyncRequest = null;
      this.setState({ trackList });
    });
  }

  onTrackSelect = track => {
    const { currentTrack, isPlaying } = this.state;

    if (currentTrack && track === currentTrack) {
      const action = isPlaying ? this.pause : this.play;
      action();
    } else {
      this.setState({
        currentTrack: track,
        isPlaying: true
      });
    }
  };

  pause = () => {
    this.state.playerRef.current.pause();
    this.setState({
      isPlaying: false
    });
  };

  play = () => {
    this.state.playerRef.current.play();
    this.setState({
      isPlaying: true
    });
  };

  render() {
    const { currentTrack, trackList, playerRef } = this.state;

    return (
      <Suspense fallback={<Loader />} maxDuration={2000}>
        <div className={styles.app}>
          <h1>Motion</h1>

          <PlayerContext.Provider value={this.state}>
            <TrackList songs={trackList} onSelect={this.onTrackSelect} />

            <Player track={currentTrack} ref={playerRef} />
          </PlayerContext.Provider>
        </div>
      </Suspense>
    );
  }
}

export default App;
