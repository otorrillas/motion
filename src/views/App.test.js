import React from 'react';
import renderer from 'react-test-renderer';

import App from './App';

describe('<App />', () => {
  describe('@render', () => {
    const renderComponent = props => renderer.create(<App {...props} />);

    it('default', () => {
      const tree = renderComponent().toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
