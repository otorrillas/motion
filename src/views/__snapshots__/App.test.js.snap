// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> @render default 1`] = `
<div
  className="app"
>
  <h1>
    Motion
  </h1>
  <table
    className="trackList"
  >
    <thead>
      <tr>
        <th />
        <th>
          Song
        </th>
        <th>
          Artist
        </th>
        <th>
          Album
        </th>
      </tr>
    </thead>
    <tbody />
  </table>
  <audio
    autoPlay={true}
    className="player"
    controls={true}
    src={null}
  />
</div>
`;
